File: apps/mobile/ios/Sources/FlagshipUI/Screens/SecretRequestsScreen.swift
import SwiftUI
import CryptoKit
import FlagshipCore
import FlagshipAPI
import Flagship

/// Phone-as-unlock-endpoint approval surface.
///
/// Opened by the `secret-request` push ("your box is finishing setup —
/// open to approve") and from a manual "pending setup" entry. On open it
/// fetches the account's pending mailbox requests, RE-VERIFIES each one
/// against the box's STK as independently resolved from the directory, and
/// shows the box's device-info for a one-tap "Yes, this is my box" confirm.
///
/// On confirm the coordinator unseals/re-seals the LUKS key (unlock-key) or
/// IRK-signs the entitlement carrier (entitlement) and posts the reply back
/// through `.com`. The box picks it up on its poll.
struct SecretRequestsContainer: View {
    @Environment(\.secretMailboxClient) private var mailbox
    @Environment(AppState.self) private var app
    @Environment(ToastCenter.self) private var toasts
    @Environment(\.colorScheme) private var scheme

    @State private var state: LoadingState<[SecretRequestCoordinator.VerifiedRequest]> = .idle
    @State private var inFlightId: String?
    private let bootUnlock = BootUnlockStore()

    var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                Text("Your box is finishing setup. Confirm it's yours to release its boot secret — only your phone can.")
                    .font(FS.font.body())
                    .foregroundColor(c.textMuted)

                switch state {
                case .idle, .loading:
                    ForEach(0..<2) { _ in ServerCardSkeleton() }
                case .failed(let msg):
                    ErrorCard(message: msg)
                case .loaded(let requests):
                    if requests.isEmpty {
                        FSCard { Text("No box is waiting for approval right now.").foregroundColor(c.textMuted) }
                    } else {
                        ForEach(requests) { req in
                            SecretRequestCard(
                                request: req,
                                isInFlight: inFlightId == req.id,
                                onApprove: { await approve(req) }
                            )
                        }
                    }
                }
            }
            .padding(FS.space.s6)
        }
        .navigationTitle("Approve box")
        .navigationBarTitleDisplayMode(.inline)
        .refreshable { await reload() }
        .task { if case .idle = state { await reload() } }
    }

    private func makeCoordinator() -> SecretRequestCoordinator? {
        guard let username = app.currentUser else { return nil }
        return SecretRequestCoordinator(
            mailbox: mailbox,
            username: username,
            irkProvider: {
                try await Keystore.deriveIRK(reason: "Approve your box's boot secret")
            },
            unsealSeedProvider: { serverDomain in
                // The installer seals the LUKS key against one of the
                // phone's Ed25519 keys. Try the per-server BAK first
                // (deterministically derivable from the UMK), then the
                // IRK — whichever it was sealed against opens it.
                var seeds: [Data] = []
                if let bak = try? await Keystore.deriveBAK(
                    serverId: serverDomain,
                    reason: "Unseal the disk key for \(serverDomain)"
                ) {
                    seeds.append(bak.rawRepresentation)
                }
                if let irk = try? await Keystore.deriveIRK(reason: "Unseal the disk key") {
                    seeds.append(irk.rawRepresentation)
                }
                return seeds
            }
        )
    }

    private func reload() async {
        guard let coord = makeCoordinator() else {
            state = .failed("Sign in to approve a box.")
            return
        }
        state = .loading
        do {
            let verified = try await coord.fetchVerifiedRequests()
            state = .loaded(verified)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    private func approve(_ req: SecretRequestCoordinator.VerifiedRequest) async {
        guard let coord = makeCoordinator() else { return }
        inFlightId = req.id
        defer { inFlightId = nil }
        do {
            // "auto" servers (the default) also get a box-sealed lease so future
            // boots self-unlock without the phone. The returned lease id is the
            // kill-switch handle — persist it per-server. ("approve" servers
            // deposit nothing; the box asks again every boot.)
            let depositLease = bootUnlock.effectiveMode(for: req.serverDomain) == .auto
            let leaseId = try await coord.confirmAndRespond(req, depositAutoLease: depositLease)
            if let leaseId { bootUnlock.setLeaseId(leaseId, for: req.serverDomain) }
            toasts.success("Approved \(req.serverDomain). Your box will pick it up.")
            await reload()
        } catch {
            toasts.error("Approval failed: \(error.localizedDescription)")
        }
    }
}

/// One pending request card: device-info backstop + one-tap confirm.
private struct SecretRequestCard: View {
    @Environment(\.colorScheme) private var scheme
    let request: SecretRequestCoordinator.VerifiedRequest
    let isInFlight: Bool
    let onApprove: () async -> Void

    private var purposeLabel: String {
        switch request.purpose {
        case .unlockKey:    return "Unlock its encrypted disk"
        case .entitlement:  return "Authorize it to serve your account"
        case .none:         return "Boot secret"
        }
    }

    var body: some View {
        let c = FSColors.scheme(scheme)
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s2) {
                Text(request.serverDomain).font(FS.font.mono()).foregroundColor(c.text)
                Text(purposeLabel).font(FS.font.caption()).foregroundColor(c.textMuted)

                if let info = request.deviceInfo {
                    VStack(alignment: .leading, spacing: 2) {
                        if let ip = info.ip { infoRow("IP", ip, c) }
                        if let region = info.region { infoRow("Region", region, c) }
                        if let os = info.os { infoRow("OS", os, c) }
                        if let host = info.hostname { infoRow("Host", host, c) }
                    }
                    .padding(.top, FS.space.s1)
                }

                Text("Is this the machine in front of you? Only approve if you recognise it.")
                    .font(FS.font.caption())
                    .foregroundColor(c.textMuted)
                    .padding(.top, FS.space.s1)

                FSPrimaryButton(
                    isInFlight ? "Signing…" : "Yes, this is my box",
                    enabled: !isInFlight,
                    block: true,
                    large: true
                ) {
                    Task { await onApprove() }
                }
                .padding(.top, FS.space.s2)
            }
        }
    }

    private func infoRow(_ label: String, _ value: String, _ c: FSColors) -> some View {
        HStack(spacing: FS.space.s2) {
            Text(label).font(FS.font.caption()).foregroundColor(c.textMuted).frame(width: 56, alignment: .leading)
            Text(value).font(FS.font.mono()).foregroundColor(c.text)
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/SecretRequestsScreen.kt
// Phone-as-unlock-endpoint approval surface (the v2 sealed-key RELAY flow).
// Kotlin mirror of FlagshipUI/Screens/SecretRequestsScreen.swift.
//
// Opened by the `secret-request` push ("your box is finishing setup — open to
// approve") and routed via DeepLink.SecretRequests. On open it fetches the
// account's pending mailbox requests, RE-VERIFIES each one against the box's
// STK as independently resolved from the directory, and shows the box's
// device-info for a one-tap "Yes, this is my box" confirm. On confirm the
// coordinator unseals/re-seals the LUKS key and posts the reply through .com;
// for "auto" servers it also deposits a box-sealed self-unlock lease (the
// returned lease id is persisted per-server for the kill switch).

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.api.DeviceInfoHint
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.LocalSecretMailboxClient
import com.flagshipserver.app.core.LocalToastCenter
import com.flagshipserver.app.core.SecretPurpose
import com.flagshipserver.app.core.SecretRequestCoordinator
import com.flagshipserver.app.core.ServerSettingsStore
import com.flagshipserver.app.keystore.Keystore
import com.flagshipserver.app.keystore.KeystoreIrkAccess
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.LoadingState
import kotlinx.coroutines.launch

@Composable
fun SecretRequestsScreen(nav: NavController) {
    val mailbox = LocalSecretMailboxClient.current
    val app = LocalAppState.current
    val toasts = LocalToastCenter.current
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val username by app.currentUser.collectAsState()
    val pods by app.pods.collectAsState()

    val store = remember { ServerSettingsStore.from(context) }
    var state by remember {
        mutableStateOf<LoadingState<List<SecretRequestCoordinator.VerifiedRequest>>>(LoadingState.Idle)
    }
    var inFlightId by remember { mutableStateOf<String?>(null) }

    val coordinator = remember(username, mailbox) {
        username?.let {
            // Slice D — the entitlement/lease orders minted in the ceremony sign
            // with the admin master root when this device holds one (transport
            // envelopes stay IRK).
            SecretRequestCoordinator(
                mailbox = mailbox,
                username = it,
                irk = KeystoreIrkAccess(),
                adminSigner = { r -> if (Keystore.hasAdminRoot()) Keystore.adminRootKey(r) else null },
            )
        }
    }

    suspend fun reload() {
        val coord = coordinator
        if (coord == null) {
            state = LoadingState.Failed("Sign in to approve a box.")
            return
        }
        state = LoadingState.Loading
        state = try {
            LoadingState.Loaded(coord.fetchVerifiedRequests())
        } catch (t: Throwable) {
            LoadingState.Failed(t.message ?: "Couldn't load requests")
        }
    }

    LaunchedEffect(coordinator) { reload() }

    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(scroll)
            .padding(horizontal = FS.space.s6),
    ) {
        Spacer(Modifier.height(FS.space.s8))
        FSGhostButton(label = "← Back", onClick = { nav.popBackStack() })
        Spacer(Modifier.height(FS.space.s3))
        Text(
            "Approve box",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Spacer(Modifier.height(FS.space.s2))
        Text(
            "Your box is finishing setup. Confirm it's yours to release its boot secret — only your phone can.",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 14.sp, lineHeight = 20.sp),
        )
        Spacer(Modifier.height(FS.space.s6))

        when (val s = state) {
            LoadingState.Idle, LoadingState.Loading -> ServerCardSkeleton()
            is LoadingState.Failed -> ErrorCard(s.message, onRetry = { scope.launch { reload() } })
            is LoadingState.Loaded -> {
                if (s.value.isEmpty()) {
                    FSCard(padding = PaddingValues(FS.space.s4)) {
                        Text(
                            "No box is waiting for approval right now.",
                            color = FS.colors.textMuted,
                            style = TextStyle(fontSize = 14.sp),
                        )
                    }
                } else {
                    s.value.forEach { req ->
                        // First boot vs established reboot: a box that has never
                        // come online (no check-in, no cert) is on its FIRST boot,
                        // where this approval also authorizes serving — show the
                        // full copy. A box that has come online before is just
                        // re-unlocking. Unknown (pod not in the list yet) defaults
                        // to first-boot = the fuller copy (today's wording).
                        val isFirstBoot = pods
                            .firstOrNull { it.fqdn.equals(req.serverDomain, ignoreCase = true) }
                            ?.cameOnline?.not() ?: true
                        SecretRequestCard(
                            req = req,
                            isFirstBoot = isFirstBoot,
                            inFlight = inFlightId == req.id,
                            onApprove = {
                                val coord = coordinator ?: return@SecretRequestCard
                                inFlightId = req.id
                                scope.launch {
                                    try {
                                        // "auto" servers (default; absent ⇒ auto) also get a
                                        // box-sealed self-unlock lease. The returned id is the
                                        // kill-switch handle — persist it per-server.
                                        val deposit = store.effectiveMode(req.serverDomain) ==
                                            ServerSettingsStore.Mode.AUTO
                                        val leaseId = coord.confirmAndRespond(req, depositAutoLease = deposit)
                                        if (leaseId != null) store.setLeaseId(req.serverDomain, leaseId)
                                        toasts.success("Approved ${req.serverDomain}. Your box will pick it up.")
                                        reload()
                                    } catch (t: Throwable) {
                                        toasts.error("Approval failed: ${t.message}")
                                    } finally {
                                        inFlightId = null
                                    }
                                }
                            },
                        )
                        Spacer(Modifier.height(FS.space.s3))
                    }
                }
            }
        }
        Spacer(Modifier.height(FS.space.s12))
    }
}

@Composable
private fun SecretRequestCard(
    req: SecretRequestCoordinator.VerifiedRequest,
    /** The box behind this request has never come online — its FIRST boot. */
    isFirstBoot: Boolean,
    inFlight: Boolean,
    onApprove: () -> Unit,
) {
    // On a FIRST boot the unlock approval ALSO deposits the box's entitlement
    // (consent to boot ⇒ consent to serve), so it both unlocks AND authorizes the
    // box to join the cloud — the fuller copy. On an established reboot the box is
    // already authorized, so the approval is purely a disk unlock; the "…join your
    // cloud" phrase is just noise. (The deposit still fires on every approval —
    // harmless on reboots — so this is copy only.)
    val purposeLabel = when (req.purpose) {
        SecretPurpose.UNLOCK_KEY ->
            if (isFirstBoot) "Unlock device and authorize it to join your cloud"
            else "Unlock device"
        SecretPurpose.ENTITLEMENT -> "Authorize device to join your cloud"
        null -> "Boot secret"
    }
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Column(verticalArrangement = Arrangement.spacedBy(FS.space.s2)) {
            Text(
                req.serverDomain,
                color = FS.colors.text,
                style = TextStyle(fontSize = 16.sp, fontFamily = FontFamily.Monospace),
            )
            Text(
                purposeLabel,
                color = FS.colors.textMuted,
                style = TextStyle(fontSize = 13.sp),
            )
            req.deviceInfo?.let { info ->
                DeviceInfoRows(info)
            }
            Text(
                "Is this the machine in front of you? Only approve if you recognise it.",
                color = FS.colors.textMuted,
                style = TextStyle(fontSize = 13.sp, lineHeight = 18.sp),
            )
            Spacer(Modifier.height(FS.space.s1))
            FSPrimaryButton(
                label = if (inFlight) "Signing…" else "Yes, this is my box",
                onClick = onApprove,
                enabled = !inFlight,
                block = true,
                large = true,
            )
        }
    }
}

@Composable
private fun DeviceInfoRows(info: DeviceInfoHint) {
    Column(verticalArrangement = Arrangement.spacedBy(FS.space.s1)) {
        info.ip?.let { InfoRow("IP", it) }
        info.region?.let { InfoRow("Region", it) }
        info.os?.let { InfoRow("OS", it) }
        info.hostname?.let { InfoRow("Host", it) }
    }
}

@Composable
private fun InfoRow(label: String, value: String) {
    Row {
        Text(
            label,
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 12.sp),
            modifier = Modifier.width(56.dp),
        )
        Text(
            value,
            color = FS.colors.text,
            style = TextStyle(fontSize = 12.sp, fontFamily = FontFamily.Monospace),
        )
    }
}
